trecs/models/mf.py

```diff
@@ -0,0 +1,293 @@
+        >>> mf.users_hat.shape
+        (1200, 2000) # <-- 1200 users, 2000 attributes
+
+        Note that all arguments passed in at initialization must be consistent -
+        otherwise, an error is thrown. For example, one cannot pass in
+        `num_users=200` but have `user_representation.shape` be `(300, 100)`.
+        Likewise, one cannot pass in `num_items=1000` but have
+        `item_representation.shape` be `(100, 500)`.
+
+        To use the model, it's advised to use the `startup_and_train` method, which
+        will collect user interaction data from random recommendations. At the end
+        of the startup period, an implicit matrix factorization model is fit to the
+        interaction data, generating new user/item latent representations. These
+        representations are fixed and used for future calls to `run`.
+
+        >>> mf = ImplicitMF(num_latent_features=10) # 10 latent features
+        >>> mf.startup_and_train(50) # 50 timesteps of training
+        >>> mf.run(100)
+
+        If you'd like to retrain the matrix factorization model at any point, you
+        may use the `fit_mf` method. It will fit a new set of latent user/item
+        representations based on the interaction data from the *most recent* call to
+        `run()`.
+
+        >>> mf.run(50) # data from this will not be used in fitting a new MF model
+        >>> mf.run(50) # data from this WILL be used in fitting a new MF model
+        >>> mf.fit_mf()
+
+    """
+
+    def __init__(  # pylint: disable-all
+        self,
+        num_users=None,
+        num_items=None,
+        num_latent_factors=None,
+        user_representation=None,
+        item_representation=None,
+        actual_user_representation=None,
+        actual_item_representation=None,
+        probabilistic_recommendations=False,
+        seed=None,
+        verbose=False,
+        num_items_per_iter=10,
+        model_params=None,
+        **kwargs
+    ):
+        # check inputs
+        num_users, num_items, num_attributes = validate_user_item_inputs(
+            num_users,
+            num_items,
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            100,
+            1250,
+            10,
+            num_latent_factors,
+        )
+        if model_params is None:
+            model_params = {}
+        num_features_vals = non_none_values(
+            model_params.get("features"), num_latent_factors, num_attributes
+        )
+        if len(num_features_vals) > 1:
+            raise ValueError("Number of latent factors is not the same across inputs.")
+
+        self.num_latent_factors = num_attributes
+        self.model_params = model_params
+        self.als_model = None  # initialize empty model
+        self.all_interactions = pd.DataFrame(columns=["user", "item"])  # empty interactions matrix
+
+        # generate user and item representations as needed
```

Comment:
I don't think these are the right defaults. Generally, latent factors tend to be floats and roughly normally distributed (if I am remembering something I have read before correctly).  I had these lines in the notebook with the fitting function, but they were definitely easy to miss. I think we should consider setting the default representation as


```
user_representation = np.random.normal(size=(num_users, num_attributes))
item_representation = np.random.normal(size=(num_attributes, num_users))
```

and potentially have a check for representations that users pass in as:
```

if not user_representation.dtype == 'float64':
    raise TypeError("user representations for MF should be floats")

if not item_representation.dtype == 'float64':
    raise TypeError("item representations for MF should be floats")
```
I haven't tinkered with it, so it may not completely matter if you start with binaries or floats, but I think this still might be reasonable

Reply:
Yeah, I was going off of how we've generated user/item representations by default in all of our other models. I agree that using a random normal is probably more reasonable for the default, although in most cases, these initial defaults will just get overwritten at the end of the startup phase, when the call to `fit_mf` happens. At the same time, I want to allow users to have the flexibility to pass in whatever kinds of latent representations they want, so I think that enforcing the type of the inputs would not be optimal here.